### PR TITLE
resolves: #2443 serialized plans

### PR DESCRIPTION
**Pull request by normen662**
- added plan serialization/deserialization framework
- added `PlanSerializable` as the root interface to implement plan serialization
- added tons of protobuf message definitions
- added way of extending proto definitions in downstream projects
- removed usages of dangerous plan fragments such as protobuf descriptors or runtime-specific data structures
- rewrote some operators to make certain planner specific attributes nullable as those are not serialized
- `DynamicEnum` ad `EnumLite` extensions to allow for serialization of plans using enums as part of their plan fragments (e.g. comparison `field = RED`)

### CODE REVIEW

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/serialization/PlanSerialization.java [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+    private static final String PREFIX = "META-INF/services/";
+
+    private static final Map<Class<? extends Message>, Method> fromProtoMapMethodMap =
+            loadFromProtoMethodMap();
+
+    /**
+     * Parse a single line from the given configuration file, adding the
+     * name on the line to set of names if not already seen.
+     */
+    private static int parseLine(BufferedReader r, int lc, Set<String> namesAlreadySeen, Set<String> names)
```
Comment (alecgrieser) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/serialization/PlanSerialization.java:76]: Why are we parsing the config files ourselves? Is this because the `ServiceLoader` returns a list of instantiated instances, and you need a collection of `Class<?>` objects instead? And the `ServiceLoader` can't load the instances because they need a default constructor to be service loadable, which this is not.

I don't love having custom parsing logic of these fields here (or the use of reflection, if I'm honest). I think the way we've typically done this before is by introducing a `Factory` class that can have default constructors, and then making the `Factory` the thing in the service loader.

For example, I think if this were the interface:

```java
public interface PlannerDeserializer<M extends Message, P extends PlannerSerializable> {
    Class<M> getProtoMessageClass();
    P fromProto(ProtoSerializationContext serializationContext, M protoMessage);
}
```

Then I think you could use a `ServiceLoader` to collect those factory implementations without the need for custom parsing logic here or reflection

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/serialization/PlanSerialization.java [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+ * Class to facilitate serialization and deserialization of
+ * {@link com.apple.foundationdb.record.query.plan.plans.RecordQueryPlan}s.
+ */
+public class PlanSerialization {
+    private static final Logger logger = LoggerFactory.getLogger(PlanSerialization.class);
+
+    private static final String PREFIX = "META-INF/services/";
+
+    private static final Map<Class<? extends Message>, Method> fromProtoMapMethodMap =
+            loadFromProtoMethodMap();
```
Comment (alecgrieser) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/serialization/PlanSerialization.java:70]: Instead of this being static, should there be some kind of configurable `PlannerDeserializerRegistry` that can be configured by the user like, for example, the `IndexMaintainerRegistry`? The benefit of that kind of approach is that it lets downstream adopters choose their own dependency injection strategy for any custom deserializers

**Review by alecgrieser — CHANGES_REQUESTED** [review]
I haven't looked at too many of the individual implementations yet, but I had a few comments on the basic framework that I figured were good to get out before taking a deeper look

File: fdb-extensions/src/main/java/com/apple/foundationdb/annotation/ProtoMessage.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+package com.apple.foundationdb.annotation;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Annotation for classes that serialize as part of a query plan.
```
[the remark this answers is not in the document]
Comment (normen662) [reply fdb-extensions/src/main/java/com/apple/foundationdb/annotation/ProtoMessage.java:29]: It has to be upstream, as the annotation processor needs it and that has to be available when for is compiled.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/Column.java [hunk 4]
```diff
@@ -89,4 +92,21 @@ public int planHash(@Nonnull final PlanHashMode hashMode) {
...
         }
     }
+
+    @Nonnull
+    @Override
+    public PColumn toProto(@Nonnull final PlanSerializationContext serializationContext) {
+        PColumn.Builder builder = PColumn.newBuilder();
+        builder.setField(field.toProto(serializationContext));
+        builder.setValue(value.toValueProto(serializationContext));
+        return builder.build();
```
[the remark this answers is not in the document]
Comment (normen662) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/Column.java:102]: agreed.

File: fdb-record-layer-core/src/main/proto/record_query_plan.proto [hunk 5]
```diff
@@ -0,0 +1,1501 @@
...
+  optional PValue value = 2;
+  optional PComparison comparison = 3;
+}
+
+//
+// Plans
+//
+message PPhysicalQuantifier {
+  optional string alias = 1;
+  repeated PPlanReference plan_references = 2;
```
[the remark this answers is not in the document]
Comment (normen662) [reply fdb-record-layer-core/src/main/proto/record_query_plan.proto:1028]: it's for the case in the future when we may want to serialize variations.

File: fdb-record-layer-core/src/main/proto/record_query_plan.proto [hunk 6]
```diff
@@ -0,0 +1,1501 @@
...
+}
+
+//
+// Plan reference holder to assign ids to plans in order to support plan DAGs.
```
[the remark this answers is not in the document]
Comment (normen662) [reply fdb-record-layer-core/src/main/proto/record_query_plan.proto:1032]: In general, we will encounter DAGs in the future. Mike even suggested adding support for cycles. The easiest way to create something like that is a self join over a view:

```
CREATE VIEW V AS SELECT * from T ORDER BY a WHERE ...
```

and then

```
SELECT * FROM V v1, V v2 WHERE ...
```

depending on the cost of computing `V` which is normally just macroed in, the planner may decide to create a common subexpression, materialize it and read it twice.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/PlanSerializationContext.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+        this(mode, IdentityBiMap.create());
+    }
+
+    public PlanSerializationContext(@Nonnull final PlanHashMode mode, @Nonnull final IdentityBiMap<RecordQueryPlan, Integer> knownPlansMap) {
+        this.mode = mode;
+        this.knownPlansMap = knownPlansMap;
+    }
+
+    @Nonnull
+    public PlanHashMode getMode() {
```
[the remark this answers is not in the document]
Comment (normen662) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/PlanSerializationContext.java:52]: `PlanHashable` and `PlanSerialization` will go hand in hand. `PlanHashMode` is versioned so we it 'll be able to directly code migration logic into the deserializer.

File: fdb-record-layer-core/src/main/proto/record_query_plan.proto [hunk 8]
```diff
@@ -0,0 +1,1352 @@
...
+  optional bool children_map_is_null = 1;
+  repeated IntChildPair child_pair = 2;
+  optional PValue value = 3;
+}
+
+//
+// Values
+//
+message PValue {
+  extensions 5000 to max;
```
[the remark this answers is not in the document]
Comment (normen662) [reply fdb-record-layer-core/src/main/proto/record_query_plan.proto:175]: I think we should talk about this.

File: fdb-record-layer-core/src/main/proto/record_query_plan.proto [hunk 9]
```diff
@@ -0,0 +1,1352 @@
...
+      SUB_FI = 34;
+      SUB_FL = 35;
+      SUB_FF = 36;
+      SUB_FD = 37;
+      SUB_DI = 38;
+      SUB_DL = 39;
+      SUB_DF = 40;
+      SUB_DD = 41;
+
+      MUL_II = 42;  // my favorite
```
[the remark this answers is not in the document]
Comment (normen662) [reply fdb-record-layer-core/src/main/proto/record_query_plan.proto:283]: Your comment is actually geared towards how that `Value` is implemented. You can encapsulate the input arguments to bring them to the max type and then perform the operation or you can have a physical operator for every possible combination. We went with the latter.

This PR does not attempt to be optimal here, we try to just reflect the implementation in proto: If the java class follows a particular structure, the proto structure will attempt to mimic that as close as possible.

File: fdb-extensions/src/main/java/com/apple/foundationdb/annotation/CheckProtoMessageAnnotationProcessor.java [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+
+    @Override
+    public synchronized void init(@Nonnull final ProcessingEnvironment processingEnv) {
+        super.init(processingEnv);
+    }
+
+    @Nonnull
+    @Override
+    public SourceVersion getSupportedSourceVersion() {
+        return SourceVersion.latestSupported();
```
[the remark this answers is not in the document]
Comment (alecgrieser) [reply fdb-extensions/src/main/java/com/apple/foundationdb/annotation/CheckProtoMessageAnnotationProcessor.java:66]: I'm not thinking of anything specific. My main concern was just that I wanted to understand what `SourceVersion` is doing here and what implications it has for changing JVMs (or Java source versions when we want to start using Java 17 language features)

File: fdb-record-layer-core/src/main/proto/record_query_plan.proto (hunk 8 above)
[the remark this answers is not in the document]
Comment (alecgrieser) [reply fdb-record-layer-core/src/main/proto/record_query_plan.proto:175]: Sure

File: fdb-record-layer-core/src/main/proto/record_query_plan.proto (hunk 9 above)
[the remark this answers is not in the document]
Comment (alecgrieser) [reply fdb-record-layer-core/src/main/proto/record_query_plan.proto:283]: I think what this PR shows is that the approach of enumerating out the various types quickly falls to the tyranny of the fundamental counting principle. I've found this pretty annoying when trying to update Java code (if I'm perfectly honest), but I think this is worse because it it bakes this choice into the serialization format which makes it harder to change in the future. If we, for example, want to introduce a new "type" here for the left and right arguments, then we'll have to introduce them without changing the enum numbers of any of the existing types.

I think this is a place where it would be better for the serialization to diverge from the POJO structure (or for the POJO structure to similarly change)

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/serialization/PlanSerialization.java (hunk 1 above)
Comment (alecgrieser) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/serialization/PlanSerialization.java:81]: Wouldn't using a `Factory` interface let us avoid that before Java 17 (and avoid reflection post-Java 17)? I think it may also let us avoid the annotation that was added in this PR, if I'm understanding the annotation's purpose correctly

File: fdb-record-layer-core/src/main/proto/record_query_plan.proto (hunk 5 above)
[the remark this answers is not in the document]
Comment (alecgrieser) [reply fdb-record-layer-core/src/main/proto/record_query_plan.proto:1028]: If that's not documented somewhere, it would probably be good to do so.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/PlanSerializationContext.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (alecgrieser) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/PlanSerializationContext.java:52]: So, if I'm understanding this correctly, the idea is to re-use the `PlanHashMode` version rather than to introduce a new `PlanSerialization` version?

File: fdb-extensions/src/main/java/com/apple/foundationdb/annotation/ProtoMessage.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (alecgrieser) [reply fdb-extensions/src/main/java/com/apple/foundationdb/annotation/ProtoMessage.java:29]: I think my preference would then be for a separate annotation processing package, rather than putting it in `fib-extensions`.

File: fdb-record-layer-core/src/main/proto/record_query_plan.proto (hunk 8 above)
[the remark this answers is not in the document]
Comment (normen662) [reply fdb-record-layer-core/src/main/proto/record_query_plan.proto:175]: I changed this as per discussion into a stub that can either take an extension or an `Any` or (if we wanted to) our own version of `Any`

File: fdb-extensions/src/main/java/com/apple/foundationdb/annotation/CheckProtoMessageAnnotationProcessor.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (normen662) [reply fdb-extensions/src/main/java/com/apple/foundationdb/annotation/CheckProtoMessageAnnotationProcessor.java:66]: Every annotation processor has to declare the supported versions. This particular return is to most lenient.

File: fdb-extensions/src/main/java/com/apple/foundationdb/annotation/ProtoMessage.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (normen662) [reply fdb-extensions/src/main/java/com/apple/foundationdb/annotation/ProtoMessage.java:29]: This is the same for another annotation processor we use. I will log an issue to make that its own package.

File: fdb-record-layer-core/src/main/proto/record_query_plan.proto (hunk 9 above)
[the remark this answers is not in the document]
Comment (normen662) [reply fdb-record-layer-core/src/main/proto/record_query_plan.proto:287]: I think I found a compromise which removes the hard-to-maintain switch statements by keeping some enum maps.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/PlanSerializationContext.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (normen662) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/PlanSerializationContext.java:66]: Yes and no -- it's up to us. I think we should have the versioning of both by in sync with one another.

File: fdb-record-layer-core/src/main/proto/record_query_plan.proto (hunk 5 above)
[the remark this answers is not in the document]
Comment (normen662) [reply fdb-record-layer-core/src/main/proto/record_query_plan.proto:984]: `ExpressionRef<>` and `GroupExpressionRef` are their counterparts. Will add a comment.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/serialization/PlanSerialization.java (hunk 2 above)
Comment (normen662) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/serialization/PlanSerialization.java]: done.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/serialization/PlanSerialization.java (hunk 1 above)
Comment (normen662) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/serialization/PlanSerialization.java]: So I played with your suggested approach quite a bit to get this as concise as possible. I had thought about that before I went the route you see in the PR but I ended up quite disliking it for the additional boilerplate it introduces. There are 200ish implementations that now have to define a static deserialization class. Quite tedious. I also started to try to get the serializer autogenerated through the `ProtoMessage` annotation but that didn't work as the generated code needed to create the `AutoService` annotation which for some reason didn't get picked up to do its magic in `META-INF`. I will try to get Mike's opinion on this. It looks as if we just need to wait until that bug in the JVM is fixed and I would be able to cut a lot of code.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/expressions/Comparisons.java [hunk 11]
```diff
@@ -174,6 +191,8 @@ private static Comparable toComparable(@Nullable Object obj) {
         } else if (obj instanceof UUID) {
             UUID uuid = (UUID)obj;
             return new UnsignedUUID(uuid.getMostSignificantBits(), uuid.getLeastSignificantBits());
+        } else if (obj instanceof Internal.EnumLite) {
+            return ((Internal.EnumLite)obj).getNumber();
```
Comment (MMcM) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/expressions/Comparisons.java:195]: I think this fixes #763. Maybe we should include the `enumFieldsWithoutIndex` test from #770 to confirm that.

**Review by MMcM — APPROVED** [review]

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/PlanSerializationContext.java [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+import com.apple.foundationdb.record.query.plan.plans.RecordQueryPlan;
+import com.apple.foundationdb.record.query.plan.serialization.DefaultPlanSerializationRegistry;
+import com.apple.foundationdb.record.query.plan.serialization.PlanSerializationRegistry;
+import com.google.common.base.Verify;
+
+import javax.annotation.Nonnull;
+import java.util.Objects;
+
+/**
+ * TBD.
```
Comment (alecgrieser) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/PlanSerializationContext.java:35]: At some point, we probably do want to expand on this. We should probably mark this class as `INTERNAL`, I think

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/keyspace/KeySpaceDirectory.java [hunk 13]
```diff
@@ -838,6 +839,7 @@ public enum KeyType {
 
         // Function that tests a value to see if it is of this type
         @Nonnull
+        @SpotBugsSuppressWarnings("SE_BAD_FIELD")
```
Comment (alecgrieser) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/keyspace/KeySpaceDirectory.java:842]: There is something definitely weird about this check failure. I'm pretty sure enums don't even serialize in a way that requires serializing the fields, so I think some wires got crossed. But I'm definitely fine with taking this suppression to avoid blocking this PR

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/rules/PushFilterThroughFetchRule.java [hunk 14]
```diff
@@ -236,11 +235,6 @@ private QueryPredicate pushLeafPredicate(@Nonnull RecordQueryFetchFromPartialRec
                                              @Nonnull CorrelationIdentifier oldInnerAlias,
                                              @Nonnull CorrelationIdentifier newInnerAlias,
                                              @Nonnull final QueryPredicate leafPredicate) {
-        if (leafPredicate instanceof QueryComponentPredicate) {
-            // We cannot push these predicates. They always contain nesteds.
-            return null;
-        }
```
Comment (alecgrieser) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/rules/PushFilterThroughFetchRule.java:242]: Is this change somehow related to serialization?

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/serialization/PlanSerialization.java [hunk 15]
```diff
@@ -0,0 +1,240 @@
...
+        Verify.verify(protoEnumClass.isEnum());
+        final E1[] javaEnumConstants = domainEnumClass.getEnumConstants();
+        final E2[] protoEnumConstants = protoEnumClass.getEnumConstants();
+        Verify.verify(javaEnumConstants.length == protoEnumConstants.length);
+
+        final EnumBiMap<E1, E2> enumBiMap = EnumBiMap.create(domainEnumClass, protoEnumClass);
+        for (int i = 0; i < javaEnumConstants.length; i ++) {
+            final E1 e1 = javaEnumConstants[i];
+            final E2 e2 = protoEnumConstants[i];
+            Verify.verify(e1.ordinal() == e2.ordinal());
```
Comment (alecgrieser) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/serialization/PlanSerialization.java:234]: Hm, do we care if the ordinals match? I can think of examples where we may have trouble keeping the ordinals aligned (e.g., if an enum value is added in the middle of a POJO definition because it groups well with existing values, but we want to put it at the end of the proto definition to simplify enum number assignment).

I'm okay with keeping this check here for now and then removing it later if/when we find it to be problematic

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/serialization/PlanSerialization.java [hunk 16]
```diff
@@ -0,0 +1,240 @@
...
+import java.util.Objects;
+import java.util.UUID;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+/**
+ * Class to facilitate serialization and deserialization of
+ * {@link com.apple.foundationdb.record.query.plan.plans.RecordQueryPlan}s.
+ */
+public class PlanSerialization {
```
Comment (alecgrieser) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/serialization/PlanSerialization.java:56]: Probably want this to be marked as `INTERNAL`

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/serialization/PlanSerializationRegistry.java [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+import com.apple.foundationdb.record.PlanDeserializer;
+import com.google.protobuf.Message;
+
+import javax.annotation.Nonnull;
+
+/**
+ * Plugin interface to be used by the plan serialization framework to dispatch serialization/deserialiation
+ * invocations properly.
+ */
+public interface PlanSerializationRegistry {
```
Comment (alecgrieser) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/serialization/PlanSerializationRegistry.java:32]: This should probably be marked as `UNSTABLE` or `EXPERIMENTAL`

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/serialization/DefaultPlanSerializationRegistry.java [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+    private static final String TYPE_URL_PREFIX = "c.a.fdb.types";
+
+    private final Map<Class<? extends Message>, PlanDeserializer<? extends Message, ?>> fromProtoClassDeserializerMap;
+    private final Map<String, Class<? extends Message>> fromProtoTypeUrlClassMap;
+
+    public DefaultPlanSerializationRegistry() {
+        final var methodMaps = loadFromProtoMethodMaps();
+        this.fromProtoClassDeserializerMap = methodMaps.getLeft();
+        this.fromProtoTypeUrlClassMap = methodMaps.getRight();
+    }
```
Comment (alecgrieser) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/serialization/DefaultPlanSerializationRegistry.java:51]: It feels like this should be private to avoid people accidentally re-initializing this class (which it looks like we may be doing during tests)

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/query/FDBRecordStoreQueryTestBase.java [hunk 19]
```diff
@@ -503,6 +512,59 @@ protected void setOptimizeForIndexFilters(boolean shouldOptimizeForIndexFilters)
...
+
+    /**
+     * Serialize the plan to bytes, parse those bytes, reconstruct, and compare the deserialized plan against the
+     * original plan.
+     * @param plan the original plan
+     * @return the deserialized and verified plan
+     */
+    @Nonnull
+    protected static RecordQueryPlan verifySerialization(@Nonnull final RecordQueryPlan plan) {
+        PlanSerializationContext serializationContext = new PlanSerializationContext(new DefaultPlanSerializationRegistry(),
```
Comment (alecgrieser) [inline fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/query/FDBRecordStoreQueryTestBase.java:549]:
```suggestion
        PlanSerializationContext serializationContext = new PlanSerializationContext(DefaultPlanSerializationRegistry.INSTANCE,
```

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/query/FDBRecordStoreQueryTestBase.java [hunk 20]
```diff
@@ -503,6 +512,59 @@ protected void setOptimizeForIndexFilters(boolean shouldOptimizeForIndexFilters)
...
+        final RecordQueryPlanProto.PRecordQueryPlan planProto = plan.toRecordQueryPlanProto(serializationContext);
+        final byte[] serializedPlan = planProto.toByteArray();
+        final RecordQueryPlanProto.PRecordQueryPlan parsedPlanProto;
+        try {
+            parsedPlanProto = RecordQueryPlanProto.PRecordQueryPlan.parseFrom(serializedPlan);
+        } catch (InvalidProtocolBufferException e) {
+            throw new RuntimeException(e);
+        }
+
+        serializationContext = new PlanSerializationContext(new DefaultPlanSerializationRegistry(), PlanHashable.CURRENT_FOR_CONTINUATION);
```
Comment (alecgrieser) [inline fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/query/FDBRecordStoreQueryTestBase.java:560]:
```suggestion
        serializationContext = new PlanSerializationContext(DefaultPlanSerializationRegistry.INSTANCE, PlanHashable.CURRENT_FOR_CONTINUATION);
```

File: fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/query/FDBRecordStoreQueryTestBase.java [hunk 21]
```diff
@@ -503,6 +512,59 @@ protected void setOptimizeForIndexFilters(boolean shouldOptimizeForIndexFilters)
...
+        final RecordQueryPlan deserializedPlan =
+                RecordQueryPlan.fromRecordQueryPlanProto(serializationContext, parsedPlanProto);
+        Assertions.assertEquals(plan.planHash(PlanHashable.CURRENT_FOR_CONTINUATION), deserializedPlan.planHash(PlanHashable.CURRENT_FOR_CONTINUATION));
+        Assertions.assertTrue(plan.structuralEquals(deserializedPlan));
+        return deserializedPlan;
```
Comment (alecgrieser) [inline fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/query/FDBRecordStoreQueryTestBase.java:565]: One interesting quirk here is that this means that we are always running the deserialized plans from Cascades. Theoretically, this means that we could face a bug where, for instance, a plan is broken unless we first serialize and deserialize it (e.g., there's some state that gets set up correctly during plan deserialization but is in an uninitialized state before). I don't really have a _great_ solution for it, though. We could randomly return the serialized or deserialized plan, though that wouldn't be repeatable. We could modify `executeQuery` so that it executes the plan both before serialization and after serialization and validates the same results get back, or something, though that's also a bit iffy.

This shouldn't be a concern for now, because the tests were already validating that the non-serialized plan works, but it could be a problem in the future

File: fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LucenePlanSerializationTest.java [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+                        COMPLEX_MULTIPLE_TEXT_INDEXES_WITH_AUTO_COMPLETE_STORED_FIELDS);
+        PlanSerializationContext planSerializationContext = PlanSerializationContext.newForCurrentMode();
+        final RecordQueryPlanProto.PPlanReference proto = planSerializationContext.toPlanReferenceProto(plan);
+        final byte[] bytes = proto.toByteArray();
+        final RecordQueryPlanProto.PPlanReference parsedProto =
+                RecordQueryPlanProto.PPlanReference.parseFrom(bytes);
+        planSerializationContext = PlanSerializationContext.newForCurrentMode();
+        final RecordQueryPlan parsedPlan = planSerializationContext.fromPlanReferenceProto(parsedProto);
+        Verify.verify(parsedPlan instanceof LuceneIndexQueryPlan);
+        // TODO proper verification as we currently do not properly round-trip;
```
Comment (alecgrieser) [inline fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/LucenePlanSerializationTest.java:72]: This is currently not as much of a practical problem as it would otherwise be because we do Lucene planning with the old planner, right?

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/rules/PushFilterThroughFetchRule.java (hunk 14 above)
Comment (normen662) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/rules/PushFilterThroughFetchRule.java:242]: I removed support for this kind of predicate which was long overdue. Didn't need serialization for it so rater removed it.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/serialization/DefaultPlanSerializationRegistry.java (hunk 18 above)
Comment (normen662) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/serialization/DefaultPlanSerializationRegistry.java:51]: done

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/serialization/PlanSerializationRegistry.java (hunk 17 above)
Comment (normen662) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/serialization/PlanSerializationRegistry.java:32]: marked them as `EXPERIMENTAL`

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/serialization/PlanSerialization.java (hunk 16 above)
Comment (normen662) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/serialization/PlanSerialization.java:56]: done

**Review by alecgrieser — APPROVED** [review]
